metals/src/main/scala/scala/meta/internal/metals/codeactions/FlatMapToForComprehensionCodeAction.scala

```diff
@@ -0,0 +1,520 @@
+package scala.meta.internal.metals.codeactions
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.Enumerator
+import scala.meta.Name
+import scala.meta.Pat
+import scala.meta.Term
+import scala.meta.Tree
+import scala.meta.inputs.Position
+import scala.meta.internal.metals.Buffers
+import scala.meta.internal.metals.CodeAction
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.parsing.Trees
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.CancelToken
+
+import org.eclipse.lsp4j.CodeActionParams
+import org.eclipse.{lsp4j => l}
+
+class FlatMapToForComprehensionCodeAction(
+    trees: Trees,
+    buffers: Buffers
+) extends CodeAction {
+  override def kind: String = l.CodeActionKind.RefactorRewrite
+
+  override def contribute(params: CodeActionParams, token: CancelToken)(implicit
+      ec: ExecutionContext
+  ): Future[Seq[l.CodeAction]] = Future {
+    val uri = params.getTextDocument().getUri()
+
+    val path = uri.toAbsolutePath
+    val range = params.getRange()
+    val maybeTree =
+      if (range.getStart == range.getEnd)
+        trees
+          .findLastEnclosingAt[Term.Apply](
+            path,
+            range.getStart(),
+            {
+              case _: Term.Apply => true
+              case _ => false
+            }
```

Comment:
To get the top most parent we can do:
```suggestion
            {
              case t: Term.Apply => parent.exists(!_.is[Term.Select])
              case _ => false
            }
```
So we should get the apply at top level.

Reply:
I want to get into this a little bit, because for example `List(1, 2, 3)` by itself is a `Term.Apply`

And let’s say we might have 

`List(1, 2, 3).sortBy(x => x).map(2 * _ ).filter(_ > 6)`

Reply:
We would need another check for that:
`parent.exists(!_.is[Term.Select]) && allowedForComprehension(t)`

This way the returned tree would be the one that is the las tin chain and exactly the one you are instested in.

Reply:
Can we add the check here?